In [48]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [49]:
path=r'C:\Users\oluwo\Documents\08-2025 Instacart Basket Analysis'

In [50]:
df_dep= pd.read_csv(os.path.join(path,'02 Data','Original Data', 'departments.csv'), index_col=False)

In [51]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [52]:
df_dep.columns

Index(['department_id', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
       '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21'],
      dtype='object')

In [53]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [54]:
df_dep_t=df_dep.T

In [55]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [56]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [57]:
new_header = df_dep_t.iloc[0] #take the first row od the df_dep_t for the header

In [58]:
new_header

0    department
Name: department_id, dtype: object

In [59]:
df_dep_t_new = df_dep_t[1:] #take the data under the header row for a new df

In [60]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [61]:
df_dep_t_new.columns = new_header #set a new header

In [62]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [63]:
data_dict = df_dep_t_new.to_dict('index')

In [64]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [65]:
df_prods= pd.read_csv(os.path.join(path,'02 Data','Original Data', 'products.csv'), index_col=False)

In [66]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [67]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [68]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [69]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [70]:
df_snacks= df_prods[df_prods['department_id']==19]

In [71]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# TASK

In [72]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [73]:
#creating a file import path 
path=r'C:\Users\oluwo\Documents\08-2025 Instacart Basket Analysis'

In [74]:
df_ords= pd.read_csv(os.path.join(path,'02 Data','Original Data', 'orders.csv'), index_col=False)

In [75]:
# drop eval_set column from orders.csv
df_ords=df_ords.drop(columns=['eval_set'])

In [76]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 2.Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [77]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

i will change the 'order_id', 'user_id' and 'order_number' to string from int64

In [78]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [79]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [80]:
df_ords['order_number'] = df_ords['order_number'].astype('str')

In [81]:
#confirm that the data type of order_id, user_id and order_number has changed.
df_ords.dtypes

order_id                   object
user_id                    object
order_number               object
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

# 3.Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [82]:
df_ords.head()

order_id user_id order_number  order_dow  order_hour_of_day  \
0  2539329       1            1          2                  8   
1  2398795       1            2          3                  7   
2   473747       1            3          3                 12   
3  2254736       1            4          4                  7   
4   431534       1            5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [83]:
#rename 'order_dow' to 'orders_day_of_the_week'
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [84]:
df_ords.head()

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1            1                   2                  8   
1  2398795       1            2                   3                  7   
2   473747       1            3                   3                 12   
3  2254736       1            4                   4                  7   
4   431534       1            5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 4.Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [85]:
#count of order_hour_of_day
df_ords['order_hour_of_day'].value_counts(dropna = True)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

## 10:00 is the busiest hour for placing orders.

# 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary

In [87]:
#the meaning of value 4 in 'department_id' column
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [88]:
print(data_dict.get('4'))

{'department': 'produce'}


## value 4 in the department_id column within the df_prods dataframe means "produce"

# 6.The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [89]:
#recalling product data
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [90]:
#create a subset for breakfast
df_prods_breakfast = df_prods.loc[df_prods['department_id'] == 14]

In [91]:
#confirm the product breakfast table
df_prods_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# 7.They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client

In [92]:
# create dinner party subset from alcohol, deli, beverages and meat/seafood depts
#alcohol, deli, beverages and meat/seafood depts are (5,20,7,12) respctively on the data_dict.
df_prods_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [94]:
df_prods_dinner_party.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

# 8.It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [95]:
#dinner party rows
df_prods_dinner_party.shape

(7650, 5)

## dinner party has a total count of 7650 rows and 5 column.

# 9.Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [100]:
#user1 information
df_user1=df_ords.loc[df_ords['user_id'] == '1']

In [101]:
df_user1

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1            1                   2                  8   
1   2398795       1            2                   3                  7   
2    473747       1            3                   3                 12   
3   2254736       1            4                   4                  7   
4    431534       1            5                   4                 15   
5   3367565       1            6                   2                  7   
6    550135       1            7                   1                  9   
7   3108588       1            8                   1                 14   
8   2295261       1            9                   1                 16   
9   2550362       1           10                   4                  8   
10  1187899       1           11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# 10.You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [102]:
#basic stats about user1 behaviour
df_user1.describe()

orders_day_of_week  order_hour_of_day  days_since_prior_order
count           11.000000          11.000000               10.000000
mean             2.636364          10.090909               19.000000
std              1.286291           3.477198                9.030811
min              1.000000           7.000000                0.000000
25%              1.500000           7.500000               14.250000
50%              3.000000           8.000000               19.500000
75%              4.000000          13.000000               26.250000
max              4.000000          16.000000               30.000000

## User 1 only shops between day 1 and day 4 of the week (Sunday and Wednesday)
## user 1 places orders between 7 AM and 4PM .

In [103]:
#export df_ords dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [104]:
#export df_dep_t_new dataframe
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))